---
title: "The Cellular Ecology of the Mouse Thymus"
author: "Mary Browning, John Wares, Andrew Park, Nancy Manley"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output: pdf_document
bibliography: thymus.bib
---

<<<<<<< HEAD
## Introduction2

Abstract

The correct organization and abundance of cell types is important for optimal organ performance, but no quantitative methods currently exists for distinguishing organizational patterns. The ability to identify recurrent or cryptic cellular patterns could help us to better understand tissue function and to make comparisons with diseased tissues. Here, in an attempt to overcome this barrier, we have developed  novel molecular and computational analyses that allow us to study the organ of our choice, the thymus, in a statistical framework. First, we identify cell type abundance and location and use this information to recreate a virtual thymus section.  We then treat cell types like species in an ecosystem, thus enabling us to directly apply quantitative approaches used in ecology to our virtual section. We use K means clustering to computationally produce regions of the thymus a priori, followed by Bray Curtis Dissimilarity Index to determine the degree of dissimilarity between thymus samples and regions of the thymus. The output generated from these analyses can then be used to better understand spatial patterns/biogeography within a healthy functional thymus and will enable us to make informative comparisons of thymic tissues under different developmental, mutant, or disease conditions.

=======
## Introduction
>>>>>>> origin/master


Within the field of organ biology, there is a lack of quantitative methods for analyzing the organization, as opposed to the composition, of organs and tissues. Healthy organs require more than the appropriate proportions of component parts; these parts must also be arranged in specific ways in order to function properly. Without quantitative methods for assessing organization, a significant aspect of organ biology is inaccessible to analysis. However, beyond essentially descriptive assays, we have few statistical tools available for interpreting organizational characteristics. In contrast, the fields of ecology and population genetics have a robust mathematical tradition, with multiple quantitative tools available for describing the natural world. Fortunately, and surprisingly, a striking comparison can be drawn between the interactions that occur within organ systems and those that occur within ecosystems.  For instance, cell types in an organ, like species in an ecosystem, are influenced by the availability of space and resources as well as by the presence or absence of other cell types. The presence or absence of one cell type may alter the functional role and dominance of another, or the overall service the community provides to the larger system [e.g., @wootton05]. Also, tissues are not typically composed with all cell types at equal abundance, which is similar to species in a community [1]; there are a few common "species", and many low abundance "species" that fit a normal distribution curve.  By comparing cell types within a biological organ to species in an ecosystem, statistical methods used by ecologists can be directly applied to biological systems. We seek to make this comparison in an effort to expand the quantitative tools/methods available to organ biologists through the use of a well characterized and respected field.  Our overall goal is the development of a cellular ecology, where we can model cellular relationships and patterns to better understand the "biogeography" of the organ we wish to study. We aim to develop novel quantitative tools, modeled after the ecological toolkit, to spatially and statistically characterize healthy biological tissues. The results from this analysis will enable further informative comparisons with tissues in states of disequilibrium, including disease or developmental states, in order to identify any recurring organizational differences that can potentially lead to disfunction of the organ of interest.

1. Krebs CJ (1998) Ecological Methodology, 2nd ed. Menlo Park, CA: Benjamin Cummings.

###John's Paragraph
First, we aim to model a healthy adult organ in order to create a baseline model that can be used for further comparisons. We will accomplish this by identifying location and abundance of different cell types.  We will then spatially and compositionally characterize the organ by applying statistical methods commonly used in ecology directly to the produced model. By analyzing the exact spatial location and co-location of component cell types across the domain of a particular tissue or organ, we may gain a better understanding of which interactions - indicated by codistribution and sufficient density - can be validated by independent approach and analysis.  To an extent, we recapitulate the spatial scales described above: we explore the differentiation of distinct habitats within the environment (biogeography: identifying regions of endemicity or dramatically shifted abundance), the colocalization of particular sub-groups of cells (community ecology), and we use the tenets of macroecology to the extent that we can generalize that cells need a particular density to be viable interactors, and to the extent that common cells have a larger overall distribution (this is not tautological, this follows from the first: you cannot be rare and widespread in a functional sense).  The co-distribution of constituent species in a community is often used to indicate interaction - whether competitive, trophic, or facilitative [@verberk11;@angel11].

Specific structural aspects of organs with strong 3-D structures, such as the lung, kidney, and other organs that undergo branching morphogenesis, have been modeled with some success (@Hartmann07, @Miura08, @Oates12). However, the cellular organization of other types of organs can be particularly difficult to assess quantitatively.  For example, endodermal glandular organ structure is often difficult to perceive, with organizational characteristics that are visible only with the use of molecular or cellular markers; thus detecting differences in organ structure depends in part on having distinct markers for known functional cellular subsets.  Examples of this type of organ include the liver, pancreas, and thymus, all of which are of high clinical importance.  Developing a quantitative approach to assessing the structure and function of these organs is thus of direct biomedical relevance.  We performed our analysis on our model organ, the thymus.  The thymus was chosen because it is an excellent example of cellular level organization, with a strong connection between organization and function.  The thymus consists of developing T cells supported by a complex cellular environment containing a variety of resident cell types, including thymic epithelial cells (TECs), dendritic cells, vasculature, and mesenchymal cells.  These cell types comprise multiple microenvironments that direct and support thymocytes to develop from immature progenitors into mature T cells that are both self-tolerant and self-restricted.  T cell development in the thymus requires interactions with the thymic microenvironments that provide signals for their survival, proliferation, and differentiation (@Manley12).  Despite their critical role in the generation of cellular immunity and the clinical importance of thymic regeneration, the composition and organization of thymic microenvironments and the mechanisms that promote their proper development and function are not fully understood, in part due to a lack of technical approaches for quantifying tissue-level properties.  Thus, the thymus has many characteristics that make it an excellent system for developing and testing quantitative modeling: a diverse cellular composition that can be identified with cell type-specific markers, regional organization that is required for maximal organ function,  genetic models with diverse effects on organ composition and function, assays for experimentally inducing organ degeneration and regeneration, and high biomedical relevance. To date, there are no statistical models of thymus organ structure and function and no established methods for generating one.  We seek to overcome this barrier by developing a model of the organization of cell types that can be used to better understand normal thymic organization and function as well as to evaluate states of disequilibrium. 

5. Manley NR, Richie ER, Blackburn CC, Condie BG, Sage J (2012) Structure and function of the thymic
microenvironment. Front Biosci 17: 2461-2477.

*Not sure if this paragraph is necessary.  What do you think? It needs improvement if we keep it.
The data obtained in this analysis will provide both a valuable resource for understanding the organization of a normal thymus, as well as a scientific framework and accessible toolkit for the analysis of mutant phenotypes, both within this project and in the field of organ biology in general. The first step of our analysis was to determine whether there are consistent patterns of organization throughout a single thymus as well as between different thymii of the same age and genotype. Next, we constructed a baseline model of a healthy wild-type mouse thymus to be used for comparisons with other thymus samples. We tested this model against a mutant to see if we could determine/quantify the degree to which mutants are significantly less structured than WT. 




 
 Cell Types Used in Analysis
 SHOULD I INCLUDE PARAFFIN ANTIBODIES?
 ----------- ---------- -------------------------
  CELL TYPE    MARKER
  
  cTEC       CD205     

  mTEC       UEA1, Claudin3
             Claudin4, K14
             K5, Claudin5 
              
  Blood vessel Claudin5, CD31 
  
  Mesenchyme  PDGFRb
  
  Dendritic  CD11c
  
  Tcells     Foxp3, CD25  
                                       
----------- ---------- -------------------------
 
### Hypotheses

I predict that the distribution and abundance of cell types is sufficient for automated clustering of thymic regions.  The organization of a wild-type, mature thymus will either organize into more discrete regions than a mutant thymus, or the regionalization will be more variable and less discrete.
Next, we will examine cellular absence/presence with respect to other cell types to see if any patterns emerge
By using a well characterized dissimilarity index (Bray Curtis Dissimilarity Index), we will determine the degree of similarity between different regions and subregions.
We will apply the above approaches to a mutant thymus in order to see if orgazinational differences can be detected.

 
## Methods

### Terminology
----------- ---------- -------------------------
  quadrat    spatial     sampled unit within
                          domain; cell types in
                          quadrats used to find
                          regions of similar
                          composition

  transect    spatial     linear array of 
                          sampled units 
                          (quadrats), may be
                          comprehensively or
                          randomly sampled
                          
  layer      tissue       actual tissue sample
                          used to assay 
                          single-cell thickness
                          composition
                          
  section               10 micron thick tissue sample
  
  virtual section       3 sections collapsed into 1 section
  
  virtual slice         3 virtual sections collapsed into 1 section
----------- ---------- -------------------------

### System and Data Collection

#### Mice
C57BL/6 male mice (N=4) at 5-6 weeks of age were purchased from Jackson Laboratories (Bar Harbor, Maine). Aire knockout (N=2) and WT male (N=2) mice at 5-6 weeks of age were obtained from Dr. Mark Anderson's lab at the University of California, San Francisco.

#### Tissue Preparation

Intact thymuses from the Anderson lab were removed and the right and left lobes were separated, placed in OCT compound, and immediately frozen before being shipped to the Manley lab. Mice thymii from Jackson Laboratories were dissected and frozen in the Manley lab.

#### Antibodies

Primary antibodies used in this work include the biotinylated hamster CD11c (Cat#, 1:200), rabbit Claudin 3 (Cat#:  , 1:200), rabbit Claudin 4 (Cat#: , 1:200), supernatant hamster CD31 (Cat#, 1:50), rat CD25 (Cat#: , 1:50), biotinylated UEA1 (Cat#: , 1:200), conjugated mouse Claudin5 (Cat#: , 1:200), rabbit Keratin5 (Cat#, 1:200), biotin rat Foxp3 (Cat#, 1:50), goat Keratin14 (Cat#, 1:200), goat PDGFRb (Cat#, 1:200), and rat CD205 (Cat#, 1:200). 
Secondary antibodies were purchased from Invitrogen.
 
#### Frozen Sample Preparation
The entire left lobe was cut into 10-micron serial sagittal sections using a Leica CM3050 S cryostat. Sections were collected on glass slides and assigned a number corresponding to their location.  The middle third of these slides were selected for IHC.  Of these middle third, 9 sections from each sample that passed quality control standards were used for IHC.  These sections were fixed in -20 degree C acetone for 20 seconds immediately prior to application of blocking solution (10% donkey serum/PBS) for 30 minutes at room temperature.

#### IHC
Primary antibodies were mixed in PBS and slides were covered with the antibody solution and incubated/stored overnight at 4C.  The slides were rinsed with PBS 2x5 minutes and the secondary antibodies were mixed in PBS (1:800) and applied to the slides for 30 minutes.  Slides were rinsed with PBS 2x5 minutes, mounted in FluorGel (EMS) and coverslipped.   Marker combinations were determined by availability,  biological and technical considerations such as anticipated abundance, distribution of cell types, and reagent compatibility (antibody species of origin). 

#### Microscopy, PTGui, and CellProfiler
Multiple sections from each sample were used in the analysis (9 sections from each sample from the Anderson Lab; 8 sections from WT2, 6 from WT1).  The sections were assessed by specific quality control criteria such as lack of section flaws, clarity of image collection, and signal to noise ratio. Sections that passed the QC measures  were imaged at 20x as tiled, overlapping quadrats on the Zeiss  microscope and reconstructed using PTGui.  The image produced was optimized in Adobe Photoshop to reduce background and increase signal intensity so that cell counts could be taken as quickly and accurately as possible in CellProfiler.  The reconstructed image from PTGui was too large to work in CellProfiler, so the size of the image was decreased by taking the image width and dividing by 4. CellProfiler was chosen because it has the capability to quickly and accurately identify cells and output locations (x and y coordinates) in an excel file.  Results from the cell counts were  manually checked for accuracy and input into R for the spatial analysis.  Cell counts from serial sagittal sections stained with different marker combinations were collapsed into a single virtual section.  Our experimental appraoch is essentially 2-D.  Although the 3-D nature of the stromal network is an important aspect of its functionality [cite], a 2-D picture will provide an informative, if not complete view of the thymus.  After all, the qualitative assays currently used are all viewed in 2D sections and are the basis of our current views of thymus organization and function.   

####Spatial Analysis with R 
Files containing the X and Y coordinates from CellProfiler were uploaded into R and reformatted (hist2D) into a community matrix.  The K-means function in R's stats package was used to cluster the data according to cell type, abundance, and location. The  Bray-Curtis function in the Vegan package was used to calculate the Bray-Curtis dissimilarity index for the different regions produced from K-means clustering.  We used the ConnComp function in the SDMTools package to calculate and produce the connected components.

```{r use a shell type command to read in any *.txt, echo=FALSE, warning=FALSE, message=FALSE, results='hide',fig.show='hide'}

setwd("~/GitHub/Thymus/Workspace") ###Set the working directory to Workspace

load("Mutant1.rdata")
load("Mutant1_150.rdata")
load("Mutant1_Rarefaction_150.rdata")
load("Mutant2.Rdata")
load("Mutant2_150.Rdata")
load("Mutant2_Rarefaction_150.Rdata")
load("WT1.Rdata")
load("WT1_150.Rdata")
load("WT1_Rarefaction_150.Rdata")
load("WT2.Rdata")
load("WT2_100.Rdata")
load("WT2_150.Rdata")
load("WT2_Rarefaction_150.Rdata")

library(gplots)

msz=200  #matrix size in hist2d
cluster=4  #number of clusters produced for K means
txtfiles=list.files(pattern="*.txt")  #loads in order all files within folder

mat<-matrix(data=NA,nrow=40000,ncol=33) #this will change with size of matrix and # of sections/cell types

toFill<-as.data.frame(mat)  #changes to data frame

toUse0<-seq(1,33,1) #uses all 33 columns
toUse1<-seq(1,33,3) #uses columns that correspond to Section1
toUse2<-seq(2,33,3) #Section2
toUse3<-seq(3,33,3) #Section3

for(i in toUse0){
  tmp = read.table(txtfiles[i], sep="\t", head=T)
  X=tmp[[1]]  #[,1]
  Y=tmp[[2]]
  my.xy<-hist2d(X,Y,nbins=c(msz,msz))
  obj <- my.xy$counts
  
  obj[1,1] = obj[1,1]-1
  obj[200,1] = obj[200,1]-1
  obj[1,200] = obj[1,200]-1
  obj[200,200]=obj[200,200]-1
  
  count<- as.vector(obj)
  toFill[,i]<-count
    
    }

for(i in toUse1){
  tmp = read.table(txtfiles[i], sep="\t", head=T)
  X=tmp[[1]]  #[,1]
  Y=tmp[[2]]
  my.xy<-hist2d(X,Y,nbins=c(msz,msz))
  obj <- my.xy$counts
  
  obj[1,1] = obj[1,1]-1
  obj[200,1] = obj[200,1]-1
  obj[1,200] = obj[1,200]-1
  obj[200,200]=obj[200,200]-1
  
  count<- as.vector(obj)
  toFill[,i]<-count
    
    }

for(i in toUse2){
  tmp = read.table(txtfiles[i], sep="\t", head=T)
  X=tmp[[1]]  #[,1]
  Y=tmp[[2]]
  my.xy<-hist2d(X,Y,nbins=c(msz,msz))
  obj <- my.xy$counts
  
  obj[1,1] = obj[1,1]-1
  obj[200,1] = obj[200,1]-1
  obj[1,200] = obj[1,200]-1
  obj[200,200]=obj[200,200]-1
  
  count<- as.vector(obj)
  toFill[,i]<-count
    
    }

for(i in toUse3){
  tmp = read.table(txtfiles[i], sep="\t", head=T)
  X=tmp[[1]]  #[,1]
  Y=tmp[[2]]
  my.xy<-hist2d(X,Y,nbins=c(msz,msz))
  obj <- my.xy$counts
  
  obj[1,1] = obj[1,1]-1
  obj[200,1] = obj[200,1]-1
  obj[1,200] = obj[1,200]-1
  obj[200,200]=obj[200,200]-1
  
  count<- as.vector(obj)
  toFill[,i]<-count
    
    }

result0<-toFill[,toUse0]
result1<-toFill[,toUse1]
result2<-toFill[,toUse2]
result3<-toFill[,toUse3]

#print(results)

```
### Identification of Scale

As thymic epithelial cells vary in size from X to X', we considered the problem of how to generate sampled regions (quadrats) that contained sufficient information about the local community of cells that these regions could be classified. In other words, we need to have lots of data in each quadrat to have statistical power for clustering; we need lots of quadrats to improve the spatial resolution. We produced a rarefaction curve for the different quadrat sizes and found that quadrat sizes from 0.00062 to 0.00247 mm^2 would provide the best spatial resolution and statistical power.

5. Quadrat Selection and Rarefaction: you need to explain how you did this, and the next chunk can include the code you used to achieve this. That way, our results can include one plot and a brief discussion about the selection of that size quadrat and what it means for inclusion of cell types in a given quadrat

``````{r WT1-rarefaction, echo=FALSE, warning=FALSE, message=FALSE, results='hide',fig.show='hide'}

####Must use total cell types. I take the community matrix and save it as a table. I open the table in Excel, use the If Then function to account for if a cell type is present. I then reformat it into a matrix, remove the space outside the section (if bin contains 0), save it as a txt file and then use the code below. 

mat <-read.table("Totalcounts200_Thymus1_2.txt", head=T, sep="\t")

# an example 10x10 matrix
#mat <- matrix(1:200, 10)
nc <- ncol(mat) # number of columns
nr <- nrow(mat) # number of rows

size <- 5 # size of the subset matrix
nmat <- 8 # number of submatrices

# sample indices of submatrices
#set.seed(1)
idxc <- sample(seq(2, nc - size), size = nmat, replace = TRUE)
idxr <- sample(seq(2, nr - size), size = nmat, replace = TRUE)

# create a list of 8 submatrices
res <- mapply(function(x, y) mat[seq(x, x + size - 1), seq(y, y + size - 1)],
              idxr, idxc, SIMPLIFY = FALSE)

# calculate the average cell values
mean(unlist(res))

####I run this multiple times and take the average of that.

```

```{r clustering, cache=TRUE, echo=FALSE, warning=FALSE, message=FALSE, results='hide'}
library (vegan)
library(labdsv)

# This step normalizes your data and is optional.
#spe.std <- decostand(spe, "normalize")  #You can also use "standardize". See 'help' for details.

# Do the k-means clustering [read 'help' for the 'kmeans' function to see what the arguments "centers" (clusters or k) and "nstart" (randomizations) mean].

spe.kmeans_Mutant1_All <- kmeans(result0, centers=cluster, nstart=1000)
spe.kmeans_Mutant1_S1 <- kmeans(result1, centers=cluster, nstart=1000)
spe.kmeans_Mutant1_S2 <- kmeans(result2, centers=cluster, nstart=1000)
spe.kmeans_Mutant1_S3 <- kmeans(result3, centers=cluster, nstart=1000)

```

```{r assigning clusters}

library(gplots) ###Call 1 set at a time; produces image of section

v0_Mutant1 = spe.kmeans_Mutant1_All$cluster
v1_Mutant1 = spe.kmeans_Mutant1_S1$cluster 
v2_Mutant1 = spe.kmeans_Mutant1_S2$cluster
v3_Mutant1 = spe.kmeans_Mutant1_S3$cluster

v0.1_Mutant1=matrix(v0_Mutant1,nrow=200,ncol=200)
v1.1_Mutant1=matrix(v1_Mutant1,nrow=200,ncol=200)
v2.1_Mutant1=matrix(v2_Mutant1,nrow=200,ncol=200)
v3.1_Mutant1=matrix(v3_Mutant1,nrow=200,ncol=200)

heatmap.2( v0.1_Mutant1, Rowv=FALSE, Colv=FALSE, dendrogram='none', cellnote=v0.1_Mutant1,
           notecol="black", trace='none', key=FALSE,lwid = c(.01,0.99),lhei = c(.01,.99),
           margins = c(0,0),col=c("red", "yellow", "blue", "green")) 

heatmap.2( v1.1_Mutant1, Rowv=FALSE, Colv=FALSE, dendrogram='none', cellnote=v1.1_Mutant1,
           notecol="black", trace='none', key=FALSE,lwid = c(.01,0.99),lhei = c(.01,.99),
           margins = c(0,0),col=c("red", "yellow", "blue", "green")) 

heatmap.2( v2.1_Mutant1, Rowv=FALSE, Colv=FALSE, dendrogram='none', cellnote=v2.1_Mutant1,
           notecol="black", trace='none', key=FALSE,lwid = c(.01,0.99),lhei = c(.01,.99),
           margins = c(0,0),col=c("red", "yellow", "blue", "green")) 

heatmap.2( v3.1_Mutant1, Rowv=FALSE, Colv=FALSE, dendrogram='none', cellnote=v3.1_Mutant1,
           notecol="black", trace='none', key=FALSE,lwid = c(.01,0.99),lhei = c(.01,.99),
           margins = c(0,0),col=c("red", "yellow", "blue", "green")) 

#change the colors to reflect the number of clusters.  Can also reorder so colors are consistent between sections.

```

```{r cluster subplots, echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.show='hide'}
my.order<-c(1,2,3,4) # define the order we want to plot panels
par(mfrow=c(2,2)) # make 4 subplots in 2x2 style
for (i in 1:max(spe.kmeans_Mutant1_All$cluster)){ 
  barplot(colSums(result0[which(spe.kmeans_Mutant1_All$cluster==i),]),main=i,ylim=c(0,12000)) 
  # we pick out desired cluster and plot
}

my.order<-c(1,2,3,4) # define the order we want to plot panels
par(mfrow=c(2,2)) # make 4 subplots in 2x2 style
for (i in 1:max(spe.kmeans_Mutant1_S1$cluster)){ 
  barplot(colSums(result1[which(spe.kmeans_Mutant1_S1$cluster==i),]),main=i,ylim=c(0,12000)) 
  # we pick out desired cluster and plot
}

my.order<-c(1,2,3,4) # define the order we want to plot panels
par(mfrow=c(2,2)) # make 4 subplots in 2x2 style
for (i in 1:max(spe.kmeans_Mutant1_S2$cluster)){ 
  barplot(colSums(result2[which(spe.kmeans_Mutant1_S2$cluster==i),]),main=i,ylim=c(0,12000)) 
  # we pick out desired cluster and plot
}

my.order<-c(1,2,3,4) # define the order we want to plot panels
par(mfrow=c(2,2)) # make 4 subplots in 2x2 style
for (i in 1:max(spe.kmeans_Mutant1_S3$cluster)){ 
  barplot(colSums(result3[which(spe.kmeans_Mutant1_S3$cluster==i),]),main=i,ylim=c(0,12000)) 
  # we pick out desired cluster and plot
}

```

```{r counts, echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.show='hide'}
#Looking at cellular relationships between clusters
#We can make a running sum of all the counts of all cell types by cluster

b1<-colSums(result0[which(spe.kmeans_Mutant1_All$cluster==1),])
b2<-colSums(result0[which(spe.kmeans_Mutant1_All$cluster==2),])
b3<-colSums(result0[which(spe.kmeans_Mutant1_All$cluster==3),])
b4<-colSums(result0[which(spe.kmeans_Mutant1_All$cluster==4),])

b1_S1<-colSums(result1[which(spe.kmeans_Mutant1_S1$cluster==1),])
b2_S1<-colSums(result1[which(spe.kmeans_Mutant1_S1$cluster==2),])
b3_S1<-colSums(result1[which(spe.kmeans_Mutant1_S1$cluster==3),])
b4_S1<-colSums(result1[which(spe.kmeans_Mutant1_S1$cluster==4),])

b1_S2<-colSums(result2[which(spe.kmeans_Mutant1_S2$cluster==1),])
b2_S2<-colSums(result2[which(spe.kmeans_Mutant1_S2$cluster==2),])
b3_S2<-colSums(result2[which(spe.kmeans_Mutant1_S2$cluster==3),])
b4_S2<-colSums(result2[which(spe.kmeans_Mutant1_S2$cluster==4),])

b1_S3<-colSums(result3[which(spe.kmeans_Mutant1_S3$cluster==1),])
b2_S3<-colSums(result3[which(spe.kmeans_Mutant1_S3$cluster==2),])
b3_S3<-colSums(result3[which(spe.kmeans_Mutant1_S3$cluster==3),])
b4_S3<-colSums(result3[which(spe.kmeans_Mutant1_S3$cluster==4),])

###and normalize these so that each cluster type has the same total amount of cells (in other words, we're ###getting the proportion of cell types in each cluster)

###```{r normalize, echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.show='hide'}
b1<-b1/sum(b1)
b2<-b2/sum(b2)
b3<-b3/sum(b3)
b4<-b4/sum(b4)

b1_S1<-b1/sum(b1_S1)
b2_S1<-b2/sum(b2_S1)
b3_S1<-b3/sum(b3_S1)
b4_S1<-b4/sum(b4_S1)

b1_S2<-b1/sum(b1_S2)
b2_S2<-b2/sum(b2_S2)
b3_S2<-b3/sum(b3_S2)
b4_S2<-b4/sum(b4_S2)

b1_S3<-b1/sum(b1_S3)
b2_S3<-b2/sum(b2_S3)
b3_S3<-b3/sum(b3_S3)
b4_S3<-b4/sum(b4_S3)

#We can look at the difference between pairs of clusters to see which cell types are most different between #clusters

par(mfrow=c(3,2))  ##All
barplot(b1-b2,main="1-2")
barplot(b1-b3,main="1-3")
barplot(b1-b4,main="1-4")
barplot(b2-b3,main="2-3")
barplot(b2-b4,main="2-4")
barplot(b3-b4,main="3-4")

par(mfrow=c(3,2))  ##All
barplot(b1_S1-b2_S1,main="1-2")
barplot(b1_S1-b3_S1,main="1-3")
barplot(b1_S1-b4_S1,main="1-4")
barplot(b2_S1-b3_S1,main="2-3")
barplot(b2_S1-b4_S1,main="2-4")
barplot(b3_S1-b4_S1,main="3-4")

par(mfrow=c(3,2))  ##All
barplot(b1_S2-b2_S2,main="1-2")
barplot(b1_S2-b3_S2,main="1-3")
barplot(b1_S2-b4_S2,main="1-4")
barplot(b2_S2-b3_S2,main="2-3")
barplot(b2_S2-b4_S2,main="2-4")
barplot(b3_S2-b4_S2,main="3-4")

par(mfrow=c(3,2))  ##All
barplot(b1_S3-b2_S3,main="1-2")
barplot(b1_S3-b3_S3,main="1-3")
barplot(b1_S3-b4_S3,main="1-4")
barplot(b2_S3-b3_S3,main="2-3")
barplot(b2_S3-b4_S3,main="2-4")
barplot(b3_S3-b4_S3,main="3-4")

The way to read the last plots is for something like "2-3", positive bars mean that cell type is more common in cluster 2 (than 3); negative bars mean the opposite. The higher the absolute number on the y-axis, the greater the difference is in that cell count between the two clusters. I think (as John suggested), that we can move ahead and be more formal with this analysis (e.g. Bray Curtis dissimilarity), but this is a useful first step.
```

## Bray Curtis

```{r braycurtis, message=FALSE, comment="", echo=FALSE, warning=FALSE, message=FALSE, results='hide'}
library(vegan) 
library(dplyr) # used for chaining and manipulation  
# http://blog.rstudio.org/2014/01/17/introducing-dplyr/
#reads as: take only cluster 1 data from 'spe', then take column sums (i.e. total cell type count) 
#then transpose (swap rows/cols needed for bray curtis calc)

df<-result0[which(spe.kmeans_Mutant1_All$cluster==1),] %>% colSums() %>% t()   # example of chaining. 
for (j in 2:4){  
  tmp<-result0[which(spe.kmeans_Mutant1_All$cluster==j),] %>% colSums() %>% t() 
  df<-rbind(df,tmp)
}

df_S1<-result1[which(spe.kmeans_Mutant1_S1$cluster==1),] %>% colSums() %>% t()
for (j in 2:4){  
  tmp_S1<-result1[which(spe.kmeans_Mutant1_S1$cluster==j),] %>% colSums() %>% t() 
  df_S1<-rbind(df_S1,tmp_S1)
}

df_S2<-result2[which(spe.kmeans_Mutant1_S2$cluster==1),] %>% colSums() %>% t()
for (j in 2:4){  
  tmp_S2<-result2[which(spe.kmeans_Mutant1_S2$cluster==j),] %>% colSums() %>% t() 
  df_S2<-rbind(df_S2,tmp_S2)
}

df_S3<-result3[which(spe.kmeans_Mutant1_S3$cluster==1),] %>% colSums() %>% t()
for (j in 2:4){  
  tmp_S3<-result3[which(spe.kmeans_Mutant1_S3$cluster==j),] %>% colSums() %>% t() 
  df_S3<-rbind(df_S3,tmp_S3)
}

#df<-apply(df,2,as.integer) %>% as.data.frame()
#df_S1<-apply(df_S1,2,as.integer) %>% as.data.frame()
#df_S2<-apply(df_S2,2,as.integer) %>% as.data.frame()
#df_S3<-apply(df_S3,2,as.integer) %>% as.data.frame()
```

```{r dendrogram}
#need same number of columns to bind them.
#colnames(df2)<-colnames(df)
colnames(df_S2)=colnames(df_S1)
colnames(df_S3)<-colnames(df_S1)

###This is where I import the other files.  Upload the other workspaces. I don't think I will be able to compare the sections with ALL

#dfTOTAL=rbind(df_S1,df_S2,df_S3)

dfWT=rbind(df_S1_WT1,df_S2_WT1, df_S3_WT1, df_S1_WT2, df_S2_WT2, df_S3_WT2)

dfTOTAL<-rbind(df_S1_WT1,df_S2_WT1, df_S3_WT1, df_S1_WT2, df_S2_WT2, df_S3_WT2, df_S1_Mutant1,
df_S2_Mutant1, df_S3_Mutant1, df_S1_Mutant2, df_S2_Mutant2, df_S3_Mutant2)

dfTOTAL_R<-rbind(df_S1_WT1_R,df_S2_WT1_R, df_S3_WT1_R, df_S1_WT2_R, df_S2_WT2_R, df_S3_WT2_R, df_S1_Mutant1_R,
df_S2_Mutant1_R, df_S3_Mutant1_R, df_S1_Mutant2_R, df_S2_Mutant2_R, df_S3_Mutant2_R)


BrayCurtis<-vegdist(dfWT,method="bray")
hc<-hclust(BrayCurtis)
plot(hc)

#save.image("Mutant1.rdata")
```

```{r, Connected Component labeling}
## Procedures

#Load in relevant data

load("WT1_100_4.rdata")
my.mat = v1.1_WT1

load("WT1_150.rdata")
my.mat = v1.1_WT1
my.mat = v2.1_WT1
my.mat = v3.1_WT1

load("WT2_150.rdata")
my.mat = v1.1_WT2
my.mat = v2.1_WT2
my.mat = v3.1_WT2
  
load("Mutant1_150.rdata")
my.mat = v1.1_Mutant1
my.mat = v2.1_Mutant1
my.mat = v3.1_Mutant1

load("Mutant2_150.rdata")
my.mat = v1.1_Mutant2
my.mat = v2.1_Mutant2
my.mat = v3.1_Mutant2


#Suggested notation (I realize this is a bit awkward - but let's start with this and revise notation based on #group feedback)

#* cluster type (labelled 1,2,3,4) = medullary (2 types), etc
#* cluster (labelled 1,2,3,...,n) = a physical assemblage of pixels of the same cluster type

#Each image is first analyzed by k-means clustering to form four cluster types, of which two are distinct parts #of the medullary region. For now, we take as given that one of these is type 2 (but we're working on automating #that so that cluster types are numbered consistently across images). First we isolate type-2 pixels in the #matrix of cluster types (`my.mat`). This means every pixel of type 2 becomes a "1" and every other pixel #becomes a "0". This new matrix is called `type.2`


type.2<-ifelse(my.mat==3,1,0)
type.3<-ifelse(my.mat==4,1,0)


#Next we run a CCL analysis on `type.2` to assign each pixel to membership of one of 1:n clusters. This is #stored in a matrix called `clus.2`


library(SDMTools)
clus.2<-ConnCompLabel(type.2)
clus.3<-ConnCompLabel(type.3)


#To visualize, we can just plot where pixels of type-2 are


image(type.2)
image(type.3)


#We can also color code them according to cluster membership

my.colors<-rainbow(length(unique(as.vector(clus.2))))
my.colors[1]<-"black"
image(clus.2,col=my.colors)

my.colors3<-rainbow(length(unique(as.vector(clus.3))))
my.colors3[1]<-"black"
image(clus.3,col=my.colors3)


### Summary statistics of clusters

#The number of clusters of a given type is just the biggest number in `clus.2`

print(max(clus.2))
print(max(clus.3))

#We can visualize the size distribution (number of pixels in clusters) with a histogram

hist(clus.2[which(clus.2>0)])
hist(clus.3[which(clus.3>0)])


#And we can also view this as a table

tbl2<-table(clus.2[which(clus.2>0)])
tbl2

tbl3<-table(clus.3[which(clus.3>0)])
tbl3


#We may also choose to ignore clusters that are very small (e.g. only one or two pixels)


tiny.clusters<-which(tbl2<4)
clus.2b<-clus.2
clus.2b[which(clus.2b%in%tiny.clusters)]<-0
image(clus.2b,col=my.colors)
print(max(clus.2b))

tiny.clusters<-which(tbl3<4)
clus.3b<-clus.3
clus.3b[which(clus.3b%in%tiny.clusters)]<-0
image(clus.3b,col=my.colors)
print(max(clus.3b))


#We can make a table of this new data (with tiny clusters removed) and look and mean and variance of cluster #size


tbl.2b<-table(clus.2b[which(clus.2b>0)])
mean.cluster.size.2<-mean(tbl.2b)
var.cluster.size.2<-var(tbl.2b)
mean.cluster.size.2
var.cluster.size.2

tbl.3b<-table(clus.3b[which(clus.3b>0)])
mean.cluster.size.3<-mean(tbl.3b)
var.cluster.size.3<-var(tbl.3b)
mean.cluster.size.3
var.cluster.size.3


#The variance to mean ratio is a simple way to consider aggregation (the tendency for pixels to be shared #unequally among clusters). A ratio <1 means clusters are all more or less the same size, a ratio of 1.0 means #the sizes are random, and a ratio >1 means they are aggregated (most pixels belonging to few clusters).


v2m.ratio.2<-var.cluster.size.2/mean.cluster.size.2
v2m.ratio.2

v2m.ratio.3<-var.cluster.size.3/mean.cluster.size.3
v2m.ratio.3


#Centroids of clusters can be calculated (and visualized) relatively easily (I may have made a minor math typo - #but the plot suggests these centroids are more or less right)


cntr.2b<-NULL
for (i in 1:length(tbl.2b)){
  idx2<-which(clus.2b==as.integer(names(tbl.2b)[i]),arr.ind=T)
  cntr.2b<-rbind(cntr.2b,c(as.integer(names(tbl.2b)[i]),apply(idx2,2,mean)))
}
image(clus.2b,col=my.colors)
for (i in 1:dim(cntr.2b)[1]){
  text(cntr.2b[i,2]/(1+dim(my.mat)[1]),cntr.2b[i,3]/(1+dim(my.mat)[1]),"*",col="white")
}

cntr.3b<-NULL
for (i in 1:length(tbl.3b)){
  idx3<-which(clus.3b==as.integer(names(tbl.3b)[i]),arr.ind=T)
  cntr.3b<-rbind(cntr.3b,c(as.integer(names(tbl.3b)[i]),apply(idx3,2,mean)))
}
image(clus.3b,col=my.colors)
for (i in 1:dim(cntr.3b)[1]){
  text(cntr.3b[i,2]/(1+dim(my.mat)[1]),cntr.3b[i,3]/(1+dim(my.mat)[1]),"*",col="white")
}

#We can look at distances between centroids as well. This is most easily done using the `raster` package


library(raster)
dist.2b<-pointDistance(cbind(cntr.2b[,2],cntr.2b[,3]),cbind(cntr.3b[,2],cntr.3b[,3]),lonlat=F,allpairs=T)
rownames(dist.2b)<-colnames(dist.2b)<-names(tbl.2b)
dist.2b[1:5,1:5]

image(dist.2b)


#Shortest distance from each centroid to nearest neighbor centroid, can be calculated


dist.2b[which(dist.2b==0)]<-NA
short.dist.2b<-apply(dist.2b,1,min,na.rm=T)
short.dist.2b

mean(short.dist.2b)
var(short.dist.2b)

```

7. also Andrew Sornborger's approach for separating out the geometry/geography of clusters., and/or using connected component labeling to identify size, centroid, distance of centroids among clusters

8. jackknifing : removing one cell type at a time

# Results

##K means clustering reproduces previously identified regions and identifies cryptic information

We used the K means clustering algorithm to identify geographical clusters of cell types based on compositional similarities and abundance. We used a range of K values to compare to previously identified regions of the thymus and to identify cryptic organization. K=3 reproduced the cortical and medullary regions, which was expected since a majority of the antibody markers correspond to cell types specifically located within these regions. We increased K to see if we could reproduce the CMJ or subcapsular region.  K=4 results in formation of the outside of the section, the cortex, and the subdivision of the medulla (which we called medulla 1 (M1) and medulla 1 (M2)), which could either represent the CMJ or a new unidentified functional region.  Increasing K further results in clusters forming throughout the medulla and eventually throughout the cortex. At higher K values, we see formation of the subcapsular region. We focused on K=4 for the remainder of the study because we were interested in the formation of 2 clusters (subdivision) in the medullary region.

TALK ABOUT NORMALIZED??

Should I mention this? Next, we produced graphs to show us how clusters were forming as well as to help us determine what cell types were driving the clustering.  The first graph showed the amount of cells per cell type in each cluster. The second compared the proportions of cells across clusters. 

##Bray Curtis Identifies Degree of Dissimilarity

We used the Bray Curtis Dissimilarity Index to determine the degree of compositional similarity between different regions within a thymus section (K=4), between different sections in the same individual thymus (N=3), and between sections in different thymuses (N=2). Bray Curtis values were arranged into a dendrogram for easier visualization. The clusters produced from Kmeans tended to group together by region (cortex with cortex, etc), with the exception of M1 for our first wild-type sample, which grouped more closely with M2.  We will need more samples to determine what this indicates, but the general trend from the results of both K means and Bray Curtis indicate that our wild-type samples are compositionally very similar.  We repeated these steps with mutant mice and saw a similar result: the mutant and wild-type clusters/samples both grouped with the regions their clusters corresponded to.  This result indicates that, compositionally, mutant and wild-type are not distinct. 
 
 Compare to ecosystem so that non-organ biologists have something to relate value to.
 
 Used to support why we selected under a certain K.
 
 Showed that both WT and mutant clusters are compositionally similar.

##Connected Component Labeling

Next, we wanted to determine if mutant and wild-type samples differ spatially/have different distributions with respect to the location of the M1 and M2 regions.  The results from the K means clustering show that for wild-type mice, M2 is inset in M1, while for mutant mice, M1 and M2 appear to be more adjacent.  We used connected component labeling to identify individual cluster components and calculated the centroid of these components.  

 Michael B. Dillencourt and Hannan Samet and Markku Tamminen (1992). "A general approach to connected-component labeling for arbitrary image representations". J. ACM.
 
# Discussion

Our approach is designed to take advantage of the cellular complexity of the thymus to provide a sensitive statistical method to measure organizational changes in organ state. We used immunostaining on sagittal serial sections of a wildtype mouse thymus in order to identify distinct cellular subsets, followed by the use of novel computational approaches to quantitatively identify known and cryptic cellular spatial relationships. These methods are designed to replace the largely descriptive histological and IHC-based analyses currently available in the field, thus providing a more accurate and unbiased view of thymus biogeography.  

It is critical to understand how general observed patterns compare across individuals of the same age and genotype. One of our goals was to determine whether organization is similar or variable throughout an individual thymus as well as between thymuses. K means clustering provided us the first step in making these comparisons. K means clustering allowed us to test the structure of the thymus in a statistical, as opposed to descriptive, framework that provided support for the number of compositionally distinct thymic regions.  Unlike descriptive approaches, K means clustering is automatic and dependent on objective criteria for sorting the functional cell types in a spatially explicit way. First, we used the K means clustering algorithm to locate/identify geographical clusters of cell types based on compositional similarities, position, and abundance.  This clustering algorithm generates a posterior probability that any subsample (quadrat) belongs to each of the different k groups, while maximizing the ratio of compositional differences between groups to compositional variation within groups. This approach has previously been successful in delineating distinct cell sub-populations in tissue samples (@Veronika09,@Cataldo09).  We compared the results to previously identified regions (through descriptive approaches) of the thymus to see if we recreated the same regions or discovered new information.  Our results supported previous discussion of the geography of the thymus. For K=3, we reproduced the cortex, medulla, and the empty blank space surrounding the thymus.  We then increased K to see if we could produce any functionally relevant information. For K=4, we saw subdivision in the medulla which could either be the CMJ or a new, previously undescribed subregion which could have functional significance. By further increasing K, we began to see further subdivision of the medulla, as well as clustering throughout the cortex.  We focused on K=4 for the rest of the analysis in order to further develop our methods.  We developed graphs which showed the number of each cell type within each cluster as well as graphs which allowed for comapriosns of cells between clusters. Although these visuals were useful for understanding how the K means clustering was being performed, the goal of this project was to move away from descriptive analyses and more towards statistical values.

7. Cataldo S, Ficarra E, Macii E (2009) Automated Discrimination of Pathological Regions in Tissue Images:
Unsupervised Clustering vs. Supervised SVM Classification. In: Fred A, Filipe J, Gamboa H, editors.
Biomedical Engineering Systems and Technologies: Springer Berlin Heidelberg. pp. 344-356.)

*I don't know about this paragraph either*We then used a range of K values to identify potentially cryptic organization, such as functional microenvironments.  The k groups produced are presumably functionally distinct, . Identifying the aggregation characteristics of particular cell types or groups of cell types will allow us to quantify and validate results from the predictive models of thymus "cellular ecology" as described above.  To an extent, as K goes up it may be that spatial weighting is important, but we choose to avoid the problems that weighting brings up; it may also be possible that some of the areas defined as K approaches 10+ are legitimate microenvironments and this will be addressed in a later paper.

In ecology, the identity and relative abundance of species can be compared across communities using various dissimilarity indicies (@Magurran04) The Bray Curtis Dissimilarity Index in particular is a well characterized statistical method used to quantify the compositional dissimilarity between two sites, based on counts at each site (@Bray57). We applied Bray Curtis to our samples to determine the degree of dissimilarity between clusters produced from K means (based on cell count) for different regions and sections in the same sample as well as between different samples. We organized the results into a dendrogram and found that clusters which corresponded to similar regions grouped together, which indicated that wild-type thymuses are compositionally similar.  

(10. Magurran AE (2004) Measuring Biological Diversity. Malden, MA: Blackwell Publishing.)
 [ Bray, J. R. and J. T. Curtis. 1957. An ordination of upland forest communities of southern Wisconsin. Ecological Monographs 27:325-349.]
 
Our goal was not only to identify whether groups of 2 or more cell types have similar relative distributions (between regions or within the same regions in different sections/samples), but also to quantify the extent to which mutants or diseased thymuses are significantly less structured than wild-type. We applied the same approaches described above to a mutant mouse thymus in order to determine if there are any detectable organizational differences compared to the wildtype. We selected the Aire knockout Mouse (@Anderson02) for several reasons.  First, the Aire knockout is a well characterized mutant that affects an individual cell population. Null mutations in the Aire gene, which marks a specific subpopulation of mTECs, cause defective negative selection and multi-organ autoimmunity.  Analyses of TEC phenotypes have been performed, and there are reports that Aire mutants have defects in mTEC composition and organization.  There is debate as to whether mutant mice double in mTEC number, particularly with respect to K14 (@Gillard07).  A recent report has also shown displaced dendritic cells in Aire mutant thymus (@Yano08,@Lei11). These results have functional consequences for different models of Aire function as either a regulator of tissue-restricted antigens in the thymus or as a key modulator of mTEC differentiation and mechanistic implications for autoimmunity in these mice.  All of these reports point to the need for more robust quantitative methods to evaluate cellular organization.  Although we know Aire is functionally important, we are interested in determining the degree to which Aire plays a role in maintaining proper organization of the thymus. The parallel in ecology is removal of a species from a community, which can lead to two distinct outcomes: the relative species abundance may not be affected if species are independently utilizing available resources, or the relative abundance of species may change dramatically, suggesting strong interactions depending on missing components [CITE].  This type of species is known as a keystone species, and its availability has strong implications on the proper functioning and survival of community it belongs to.  

The results from both the K means clustering algorithm and the Bray Curtis Dissimilarity Index failed to detect any compositional differences between the mutant and wildtype samples. We then checked for organizational differences since there was variation in location of M1 relative to M2. We used Connected Component Labeling (cite?) to group connected components within the same clusters together.  We were then able to calculate the centrod of each cluster and calculate the distance between clusters of different regions.  For the wild-type mouse thymus, we expected the M1 and M2 centroids to overlap since M1 surrounds M2.  For the mutant sample, we expected to see a larger distance between the centroids since the M1 and M2 clusters were adjacent to each other.

SUMMARY-



2. We believe we have identified subregions in the medulla. This is cool.
3. Though our replication is minimal, we can say that there is/is not quantifiable distinction between the spatial organization of WT and Aire- mice. This has been debated in the literature and we resolve it.
4. There is much more we can do with this approach. 
5. Ecologically, it is unusual to have the problem of such complete sampling of an ecosystem. Our approach to sampling this has strengths and concerns, generally associated with K-means clustering which has its subjective/ambiguous problems but is the best we can do!


Why this approach is novel 
Coolest thing we found, how to explain it
If we did not find a big shift in organization between treatment, it is probably because what was turned off in mutant was not a 'foundational species' in this system (see http://bioscience.oxfordjournals.org/content/61/10/782.full ) as such is a minor perturbation for organization of system, even if ecosystem service ie functionality of thymus is affected. We have to separate these effects after all! It is also important to note that the cell types are not independent arrivals in the thymic ecosystem. As a developmental process, the analogy is not complete: this would be like a coral reef developing with functionally distinct polyps spatially arranged to improve net fitness (i.e. reproductive capacity); this of course does happen in certain colonial species (hydrozoans for example) and as such this may generate its own numerical/mathematical patterns distinct from the standards of community ecology (e.g. species-area hypothesis).
Second coolest thing
Concern we would like to address

Next step for application of this approach - also bigger picture stuff.
Comparison - a manipulated salt marsh ecosystem and how B-C was used to compare treatment and sites
http://www.mass.gov/eea/docs/czm/habitat/wetlands/cape-cod-report-v2.pdf
another
http://books.google.com/books?id=Kpi9TjDQO8QC&pg=PA28&lpg=PA28&dq=bray-curtis+salt+marsh&source=bl&ots=iiISZsY9Jf&sig=YGSLp0cG06oKxB1Ung56AS2EQgw&hl=en&sa=X&ei=M44yVP-_CYylyASMioLgCQ&ved=0CDgQ6AEwAw#v=onepage&q=bray-curtis%20salt%20marsh&f=false
useful comparison because of similar taxonomic diveristy (?)
http://www.scielo.br/scielo.php?pid=S1516-89132009000600013&script=sci_arttext

http://www.researchgate.net/publication/259146839_Dissimilarity_in_plant_species_diversity_between_salt_marsh_and_neighboring_environments_decreases_as_environmental_harshness_increases

http://www.tandfonline.com/doi/full/10.1080/19425120.2014.893467#.VDKQQkuoVss
there are many more. we can learn from the methods..

# Literature Cited

assume we have vector (e.g. 20000 pixels, numbered)


# Future Directions

The next step of this project will be to examine how removal of specific cell types from the dataset will alter clustering as well as the ecological indicies.  We are also interested in further examining the clusters that appear when K is increased.  These clusters may have functional significance (for instance, they could be microenvironments that have not previously been identified).  We are also interested in looking at the process of involution, and how the organization differs from that of a WT thymus. We could also look at other mutants.
The methods utilized in this study can be adapted to other organ systems.
